### PR TITLE
Problem (WIP #1703): incorrect tree hashing in mls modules

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- included proposed fixes (5ee9f5d) to the tree math draft
- explicit types for node and leaf indices

### CODE REVIEW

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    let lvl = level(x);
+    if lvl == 0 {
+        None
+    } else {
+        let nodes = node_width(leafs);
+        // 01x -> 10x
+        let mut r = x ^ (0b11 << (lvl - 1));
+        while r >= nodes {
+            // TODO proof
+            r = left(r).unwrap();
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree_math.rs:92]: or `r = left(r)?;`

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+#[inline]
+pub fn parent(x: NodeSize, leafs: LeafSize) -> Option<NodeSize> {
+    let nodes = node_width(leafs);
+    assert!(x < nodes);
+    if x == root(leafs) {
+        return None;
+    }
+    let mut p = parent_step(x);
+    while p >= nodes {
+        p = parent_step(p)
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree_math.rs:185]: works without a semicolon at the end?

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+/// The direct path of a node, ordered from the root
+/// down, including the root
+#[inline]
+pub fn direct_path(node_pos: NodeSize, leafs: LeafSize) -> Vec<NodeSize> {
+    let mut d = Vec::new();
+    if let Some(mut p) = parent(node_pos, leafs) {
+        let r = root(leafs);
+        while p != r {
+            d.push(p);
+            p = parent(p, leafs).unwrap();
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree_math.rs:199]: proof / no panic?

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+/// The level of a node in the tree. Leaves are level 0, their parents are
+/// level 1, etc. If a node's children are at different levels, then its
+/// level is the max level of its children plus one.
+#[inline]
+pub fn level(x: NodeSize) -> u32 {
+    #[cfg(feature = "nightly")]
+    let n = x.trailing_ones();
+    #[cfg(not(feature = "nightly"))]
+    let n = {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: I kept the simpler version in the comment, we can use that when the feature become stable. it's also more efficient using special instructions.

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        if x == 0 {
+            return 0;
+        }
+
+        let mut k = 0;
+        while (x >> k) > 0 {
+            k += 1
+        }
+        k - 1
+    };
```
[the remark this answers is not in the document]
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: I kept the simpler version in the comment, we can use that when the feature become stable. it's also more efficient using special instructions.

File: chain-tx-enclave-next/mls/src/tree_math.rs [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+//!     /    \
+//!   P00X  P10X
+//!   ```
+//!   - `P` is a shared prefix
+//!   - `X` is a sequence of ones
+//!
+//!   So moving from parent to children or vice versa, is only about switching one or two bits.
+//!
+//! The left sub-tree of the root node is always complete, but the right part might not.
+type NodeSize = usize;
```
[the remark this answers is not in the document]
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: Changed to newtype and methods.

File: chain-tx-enclave-next/mls/src/tree.rs [hunk 7]
```diff
@@ -158,117 +159,6 @@ pub struct Tree {
...
-        p = parent(p, n);
-    }
-    if node_pos != r {
-        d.push(r);
-    }
-    d
-}
-
 impl Tree {
     pub fn get_package(&self, leaf_index: usize) -> Option<&KeyPackage> {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree.rs]: Done

File: chain-tx-enclave-next/mls/src/tree_math.rs (hunk 1 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: Done

File: chain-tx-enclave-next/mls/src/tree_math.rs (hunk 2 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: It seems to work without it, but still I added the semicolon.

File: chain-tx-enclave-next/mls/src/tree_math.rs (hunk 3 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree_math.rs]: Added no panic comment

**Pull request merged**